MPT: Minimal RLP encoding + Lookup improvements

Brechtpd opened the pull request:
### Description

[_PR description_]

### Issue Link

[_link issue here_]

### Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

### Contents

- [_item_]

### Rationale

[_design decisions and extended information_]

### How Has This Been Tested?

[_explanation_]

<hr>

## How to fill a PR description 

Please give a concise description of your PR.

The target readers could be future developers, reviewers, and auditors. By reading your description, they should easily understand the changes proposed in this pull request.

MUST: Reference the issue to resolve

### Single responsability

Is RECOMMENDED to create single responsibility commits, but not mandatory.

Anyway, you MUST enumerate the changes in a unitary way, e.g.

```
This PR contains:
- Cleanup of xxxx, yyyy
- Changed xxxx to yyyy in order to bla bla
- Added xxxx function to ...
- Refactored ....
```

### Design choices

RECOMMENDED to:
- What types of design choices did you face?
- What decisions you have made?
- Any valuable information that could help reviewers to think critically

adria0 committed:
Fix testool manifest and add test it in github actions (#1535)

### Description

Testool failed to compile after
https://github.com/privacy-scaling-explorations/zkevm-circuits/pull/1524,
this is due:

- testool compilation is not checked in github actions
- some functions required are not only available without `test-circuits`
feature

### Type of change

Bug fix (non-breaking change which fixes an issue)

### Contents

- Fixes Cargo.toml
- Adds basic test of testool in github actions

---------

Co-authored-by: adria0 <adria0@nowhere>

ed255 committed:
Clean up github workflows (#1537)

### Description

- Remove bitrot task which was building the benchmarks and examples but
we don't have any
- Merge cargo fmt, clippy and doc links into a single job to reduce
noise
- Split heavy tests and light tests into two jobs to parallelize the
work
- Merge all feature tests into a single job to reduce noise.
- Disable integration tests on pull requests because they are taking a
long time currently. Enable them only when the trigger-integration-test
label is set, when they are scheduled via cron, or when pushing to main.
---
- Times before: 57 minutes
  - Longest: Integration tests: 57 min
  - Second longest: Tests: 28 min
- Times now: 19 minutes
  - Longest: Light unit tests: 19 min
  - Second longest: Heavy unit tests: 16 min

ChihChengLiang committed:
Remove the prover bench from Halo2 CI script (#1538)

### Description

The CI script was introduced in #330 but was rarely used recently.

The CI script must work with the branch named
"[halo2benchmarking](https://github.com/privacy-scaling-explorations/zkevm-circuits/tree/halo2benchmarking)",
which is long gone.

We remove the related scripts to simplify the codebase.

ntampakas committed:
Testool migration (#1540)

### Description
This PR introduces testool nightly builds in GH Actions. The workflow
can also be triggered manually selecting between light and nightly
suites.
The reports are public now under:
http://testool-public.s3-website.eu-central-1.amazonaws.com/

### Issue Link

https://github.com/privacy-scaling-explorations/zkevm-circuits/issues/1505


### Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing
functionality to not work as expected)
- [ ] This change requires a documentation update

### How Has This Been Tested?
Executed manually and automatically the workflow, triggering both
suites.

---------

Co-authored-by: Chih Cheng Liang <chihchengliang@gmail.com>

ChihChengLiang committed:
Bytecode, and a collection of bytecodes. (#1529)

### Description

We revisit the handling of the bytecodes in the codebase. We reduce
manual manipulation of bytes and use a more streamlined API with
eth_types::Bytecode and bus_mapping::CodeDB.

### Issue Link

Missed out work in #1391 

### Type of change

Refactor

### Contents

- Removed witness::bytecode module.
- Extend eth_types::Bytecode and bus_mapping::CodeDB.
  - eth_types::Bytecode handles a single bytecode instance, and
  - CodeDB handles a group of bytecode instances
- rewrite the bytecode circuit witness assignment.
- rewrite the bytecode circuit tests, so that overwriting will not be
included in the production code.
- deduplicated the Mock account testing code.

### Rationale

- I kept the scope of the change small to reduce the review burden.
- I will add more rationale with review comments.

ed255 committed:
Hardcode build target in workflow (#1544)

### Description

Remove the matrix strategy in the build job and instead hardcode the
build target. Since we only have 1 target we don't need the matrix
strategy.

This is a small PR aimed to clean up the output of jobs running from
commits merged to main. Our configuration skips jobs that have already
run, and when this happens, the matrix strategy is not resolved and we
see this:

![image](https://github.com/privacy-scaling-explorations/zkevm-circuits/assets/24987329/e90fc7ae-841b-4adb-8956-5b4fa6539fe4)

By removing the matrix strategy we should see a cleaner output.

Commit:
Fully implement PUSHn (#1527)

### Description
We intend here to support PUSHn with bytecode length inferior to n.

### Issue Link
Issue #633 
Linked to specs issue
https://github.com/privacy-scaling-explorations/zkevm-specs/issues/463

### Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing
functionality to not work as expected)
- [ ] This change requires a documentation update

### Contents

- Update push.rs to support bytecode of size smaller than n
- Update stop.rs similarly

### Rationale

In pushn, "selectors" are switches to enable or disable lookup arguments
on the bytcode bytes as well as to check whether these are non-null if
needs be. They directly depend on "n" of PUSHn.
We add to these "selectors", "counters" which have similar role but are
a subset of them depending on the bytecode length. We added these extra
switches to be able to easily check the various identities at the risk
of some redundancy.

To support these changes, modifications in stop.py also had to be done.
More precisely, the check "is_out_of_range had to be extended to support
small bytecodes.

### How Has This Been Tested?

Extra tests have been added in push.rs (only).

DreamWuGit committed:
implement oog code store and max code size exceeds error state (#1545)

### Description

This PR aims to implement oog code store and max code size exceeds error
state, merge the two error cases into one gadget.
spec refer to
https://github.com/privacy-scaling-explorations/zkevm-specs/pull/475
### Issue Link

issue #1259 

### Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing
functionality to not work as expected)
- [ ] This change requires a documentation update

### Contents

- [_item_]

### Rationale

[_design decisions and extended information_]

### How Has This Been Tested?

add new test for these two error cases



This PR contains:
- add buss mapping part when  such two errors happens
- evm gadget to constraints
- Added corresponding tests

AronisAt79 committed:
Worst case benches - deploy wc contracts (#1550)

### Description

This is the first of a PRs series aiming to implement worst case block
proof bench-marking tests within zkevm-circuits/integration-tests.
The changes submitted here implement the blockchain data generation
functionality* needed for the wc bencmarks.

* compilation and deployment of new SCs

### Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
✅ New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing
functionality to not work as expected)
- [ ] This change requires a documentation update

### Contents

~~1. wcerrors.rs : Recreate the solc::SolcError and solc::SolcIoError
types with necessary methods and private functions
2. worst_case.rs: Recreate ethers::solc::Solc with necessary methods~~
3. gen_blockchain_data.rs

### Rationale

Smart contracts written for generation of blocks with increased proof
computational intensity use nested opcode calls in the inline assembly
block that violate serde recursion limits. ~~Modules in Contents section
1 and 2 are created to manipulate the compilation behavior and disable
recursion depth limits during deserialization~~

Error handling has also been added in gen_blockchain_data.rs in order to
catch solc compilation errors hidden in the CompilerOutput struct

### How Has This Been Tested?


~~https://github.com/privacy-scaling-explorations/zkevm-circuits/actions/runs/5749057365/job/15583140857~~

https://github.com/privacy-scaling-explorations/zkevm-circuits/actions/runs/5753344249/job/15596341036

---------

Co-authored-by: Chih Cheng Liang <chihchengliang@gmail.com>

ChihChengLiang committed:
Remove assembly in geth-utils (#1553)

### Description

Remove geth-utils and examples.

### Type of change

New feature (non-breaking change which adds functionality)

### Rationale

- Assembly was introduced in #91 to build examples to use geth-util. It
was not used elsewhere in the codebase.
- The examples are removed as we currently don't use geth-util like the
way the examples show. (Note that the examples also fail to run now. The
transactions in the examples would fail because the trace config
defaults to a 0 block gas limit. It runs after adding a block gas
limit.)
- I found those examples unused when reviewing how we use the tracing
API. It is much easier to review the tracing code after removing the
Assembly.
- The Readme mentioned a TODO CLI, which we never built. I think it
might be interesting to create a good first issue for people to extend
testool(https://github.com/privacy-scaling-explorations/zkevm-circuits/blob/690b2f143a94b7474f0c99493a99b0f2c53d93b8/testool/README.md#L38)
for it.

leolara committed:
Keccak refactoring with new Cell Manager (#1541)

### Description

Keccak refactoring with new Cell Manager

### Issue Link


https://github.com/privacy-scaling-explorations/zkevm-circuits/issues/1203

### Type of change

- [X] Refactoring

### Contents

- Keccak refactoring with new Cell Manager

### Rationale

See in issue

### How Has This Been Tested?

Passes unit tests.

DreamWuGit committed:
implement oog account access (#1555)

### Description

implement oog account access error state 
found spec by @KimiWu123
https://github.com/privacy-scaling-explorations/zkevm-specs/pull/476
thanks!
### Issue Link

[_link issue here_]

### Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing
functionality to not work as expected)
- [ ] This change requires a documentation update


This PR contains:
- buss mapping generates step operations
- circuit gadget with tests

miha-stopar committed:
MPT (#398)

This PR aims to provide the first version of Merkle Patricia Trie
circuit. The witnesses are currently generated
[here](https://github.com/miha-stopar/mpt). Lookups in the Keccak
circuit are simulated using internal lookup table.

This branch replaces
https://github.com/appliedzkp/zkevm-circuits/commits/merkle-patricia-trie
(didn't rebase as it seems cleaner this way, will leave the old branch
for a while before deleting it).

---------

Signed-off-by: smtmfft <smtm@taiko.xzy>
Signed-off-by: smtmfft <smtm@taiko.xyz>
Co-authored-by: Brecht Devos <Brechtp.Devos@gmail.com>
Co-authored-by: pinkiebell <40266861+pinkiebell@users.noreply.github.com>
Co-authored-by: Haichen Shen <shenhaichen@gmail.com>
Co-authored-by: z2trillion <z2trillion@users.noreply.github.com>
Co-authored-by: Eduard S <eduardsanou@posteo.net>
Co-authored-by: adria0.eth <5526331+adria0@users.noreply.github.com>
Co-authored-by: Han <tinghan0110@gmail.com>
Co-authored-by: smtmfft <99081233+smtmfft@users.noreply.github.com>
Co-authored-by: smtmfft <smtm@taiko.xzy>
Co-authored-by: Chih Cheng Liang <chihchengliang@gmail.com>
Co-authored-by: DreamWuGit <wwuwwei@126.com>
Co-authored-by: Rohit Narurkar <rohit.narurkar@protonmail.com>
Co-authored-by: Zhang Zhuo <mycinbrin@gmail.com>
Co-authored-by: Ming <hero78119@gmail.com>
Co-authored-by: Wu Sung Ming <wusm@sea.com>
Co-authored-by: xiaodino <qinyi@taikocha.in>
Co-authored-by: Mason Liang <mason@0x.org>
Co-authored-by: Steven <asongala@163.com>
Co-authored-by: Carlos Pérez <37264926+CPerezz@users.noreply.github.com>
Co-authored-by: Paul <108982045+ChengYueJia@users.noreply.github.com>
Co-authored-by: AronisAt79 <45406504+AronisAt79@users.noreply.github.com>
Co-authored-by: Mason Liang <mason@scroll.io>
Co-authored-by: JohnWick2ETH <104916412+JohnWick2ETH@users.noreply.github.com>
Co-authored-by: kunxian xia <xiakunxian130@gmail.com>
Co-authored-by: Leo Lara <leolara@users.noreply.github.com>
Co-authored-by: zengzengzenghuy <85567868+zengzengzenghuy@users.noreply.github.com>
Co-authored-by: john xu <john@taiko.xyz>
Co-authored-by: aguzmant103 <67167307+aguzmant103@users.noreply.github.com>
Co-authored-by: Richord <rzhang139@gmail.com>
Co-authored-by: Kimi Wu <leonhartwu@gmail.com>
Co-authored-by: Mickey <hedgefund996@gmail.com>
Co-authored-by: Luozhu <70309026+LuozhuZhang@users.noreply.github.com>
Co-authored-by: DoHoon Kim <59155248+DoHoonKim8@users.noreply.github.com>
Co-authored-by: jeong0982 <jeong0982@kaist.ac.kr>
Co-authored-by: Andy Arditi <andyrdt@gmail.com>
Co-authored-by: David Nevado <davidnevadoc@users.noreply.github.com>
Co-authored-by: omahs <73983677+omahs@users.noreply.github.com>
Co-authored-by: Richord <rrzhang@usc.edu>
Co-authored-by: ntampakas <nick@gmx.co.uk>
Co-authored-by: ashWhiteHat <phantomofrotten@gmail.com>
Co-authored-by: naure <naure@users.noreply.github.com>
Co-authored-by: Aurélien Nicolas <info@nau.re>
Co-authored-by: testuser-at-490752553772 <m_aronis@hotmail.com>
Co-authored-by: xiaodino <ruby@taiko.xyz>
Co-authored-by: Thomas Pan <43927782+thomaspanf@users.noreply.github.com>
Co-authored-by: Akase Cho <lightsing@users.noreply.github.com>
Co-authored-by: Vu <vu.voth@gmail.com>
Co-authored-by: Alex Beregszaszi <alex@rtfs.hu>
Co-authored-by: Enrico Bottazzi <85900164+enricobottazzi@users.noreply.github.com>
Co-authored-by: adria0 <adria0@nowhere>
Co-authored-by: Cecilia Zhang <zhangyixin319@gmail.com>
Co-authored-by: CeciliaZ030 <45245961+CeciliaZ030@users.noreply.github.com>

dyxushuai committed:
fix:  misused `max_priority_fee_per_gas` and `max_fee_per_gas` (#1559)

## Description

According to the description of
[EIP-1559](https://eips.ethereum.org/EIPS/eip-1559) and the definition
in Geth

https://github.com/ethereum/go-ethereum/blob/67979022aa8ea6a96b618c086f60b49af7d2a568/core/types/tx_dynamic_fee.go#L29-L30
```go
        GasTipCap  *big.Int // a.k.a. maxPriorityFeePerGas
        GasFeeCap  *big.Int // a.k.a. maxFeePerGas
```
So, the origin codes misused the mapping

 
## Issue Link

No

## Type of change

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing
functionality to not work as expected)
- [ ] This change requires a documentation update

---------

Co-authored-by: Chih Cheng Liang <chihchengliang@gmail.com>

DreamWuGit committed:
implement oog sha3 error state (#1558)

### Description

implement oog sha3 error state

### Issue Link



### Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing
functionality to not work as expected)
- [ ] This change requires a documentation update


This PR contains:
- refactor memory address gadget to handle overflow condition
- add oog sha3 gadget
- add tests

dyxushuai committed:
fix: remove unnecessary match (#1561)

### Description

We don't need to match `OpcodeId::PUSH32` again, because it has been
contained in previous code.

https://github.com/privacy-scaling-explorations/zkevm-circuits/blob/b36da2a4047f8811990c6d928bb1e1864f88328b/zkevm-circuits/src/evm_circuit/step.rs#L215-L216

https://github.com/privacy-scaling-explorations/zkevm-circuits/blob/b36da2a4047f8811990c6d928bb1e1864f88328b/eth-types/src/evm_types/opcode_ids.rs#L319-L321

The `is_push` contains the `PUSH32` opcode. So, let's remove the
duplicate match.

### Issue Link

No

### Type of change

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing
functionality to not work as expected)
- [ ] This change requires a documentation update

silathdiir committed:
Implement `PUSH0` instruction (#1463)

### Description

1. Add `PUSH0` opcode to eth-types. It could be parsed from `0x5f` or a
string of PUSH0.

2. In eth-types, update `is_push` function to return true for `PUSH0 ..
PUSH32`, and add a new function `is_push_with_data`, it returns true for
`PUSH1 .. PUSH32`.

3. Small fixes to replace `PUSH1 - 1` with PUSH0 value.

4. Add PUSH0 implementation to bus-mapping, and `PushGadget` to support
PUSH0 in zkevm-circuits.

### Issue Link

Close
https://github.com/privacy-scaling-explorations/zkevm-circuits/issues/1473

Reference previous PR
https://github.com/privacy-scaling-explorations/zkevm-circuits/pull/1361

zkevm-specs PR
https://github.com/privacy-scaling-explorations/zkevm-specs/pull/471

### Type of change

- [x] New feature (non-breaking change which adds functionality)

---------

Co-authored-by: Chih Cheng Liang <chihchengliang@gmail.com>

KimiWu123 committed:
Feat/#1518 ErrorOutOfGasStaticMemoryExpansion error state (#1560)

### Description

This PR was pulled from Scroll team
(https://github.com/scroll-tech/zkevm-circuits/pull/327) and 1.
supported uint64 overflow and fitted to our word lo/hi design.

### Issue Link

closed #1518

### Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing
functionality to not work as expected)
- [ ] This change requires a documentation update

---------

Co-authored-by: Akase Cho <lightsing@users.noreply.github.com>
Co-authored-by: adria0.eth <5526331+adria0@users.noreply.github.com>

KimiWu123 committed:
feat/#1576 refactor rw counter delta for CommonErrorGadget (#1577)

### Description

refactor  rw counter delta for CommonErrorGadget

### Issue Link

#1576 

### Type of change

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing
functionality to not work as expected)
- [ ] This change requires a documentation update

CeciliaZ030 commented on zkevm-circuits/src/circuit_tools/constraint_builder.rs:
```diff
@@ -1177,184 +1071,62 @@ macro_rules! _require {
-        $cb.store_dynamic_table(
-            description,
+            Box::leak($descr.into_boxed_str()),
             $tag,
```
Forgot to delete `$tag` here?

CeciliaZ030 replied:
```suggestion
```

CeciliaZ030 commented on zkevm-circuits/src/circuit_tools/constraint_builder.rs:
```diff
@@ -278,288 +360,132 @@ impl<F: Field, C: CellType> ConstraintBuilder<F, C> {
-        } else {
-            self.dynamic_lookups.insert(tag, vec![lookup]);
-        }
+        self.store_lookup_table(description, table_type, values);
```
What's the point of wrapping one more function here?

Brechtpd replied:
None it seems, I thought the function was still used elsewhere as well...

CeciliaZ030 commented on zkevm-circuits/src/mpt_circuit.rs:
```diff
@@ -238,79 +229,67 @@ impl<F: Field> MPTConfig<F> {
+                require!(@MptTableType::Mult => mult_table.table_exprs(meta));
+            });
+            vec![0.expr()]
+        });
```
I still advocate for a `proload_table()` API cuz it looks cleaner. Like you said, ideally one `meta.create_gate` is preferred.

Brechtpd replied:
It is possible to just call `store_table()` directly which I did for a long time, but I'm conflicted because currently `store_table` uses the active condition in the constraint builder, and so I think it's nice that all circuit code that is part of the main program flow is written inside a ` circuit!` to make that clear. But I also do see the point of a function that doesn't do that and indeed just stores the table columns as is in the table without the active condition applied, but then there's two methods to store tables which I wanted to avoid as well. I guess another function `load_table()` that doesn't apply the current condition could be useful and worthwhile.

CeciliaZ030 replied:
`load_table` LGTM! it should be the standard workflow for fixed table

CeciliaZ030 commented on zkevm-circuits/src/circuit_tools/memory.rs:
```diff
@@ -6,77 +6,89 @@ use halo2_proofs::{
-        self.banks.last().unwrap()
+        phase: u8,
+    ) -> &MB {
+        let table_tag = self.allocate_tag();
```
If we only ask for one `tag` for each RW pair, why don't we just use the same tag for both side? The lookup is initiated immediately in `MB::new(..)` anyways, so there's no need for extra tags right? Or maybe I'm missing its other usage?

Brechtpd replied:
We still have to have two to differentiate on the cell manager side in which column to store the inputs (loading) and where to store the table values (stores) right?

CeciliaZ030 replied:
In this case i feel like it doesn't use the tag at all tho, cuz you can 1) [init two cols with the same tag](https://github.com/CeciliaZ030/zkevm-circuits/blob/813426ba75413e49463b1b6fdad925655fe3395a/zkevm-circuits/src/circuit_tools/memory.rs#L182) 2) [adds it as (data expressions) -> (table expressions)](https://github.com/CeciliaZ030/zkevm-circuits/blob/813426ba75413e49463b1b6fdad925655fe3395a/zkevm-circuits/src/circuit_tools/constraint_builder.rs#L447) 3) [build lookup which performs meta.lookup_any directly on the stored expressions](https://github.com/CeciliaZ030/zkevm-circuits/blob/813426ba75413e49463b1b6fdad925655fe3395a/zkevm-circuits/src/circuit_tools/constraint_builder.rs#L371). For the CM it just needs to know these two cols are for this specific MB that's fine right?

Brechtpd replied:
You still have to load and store to the different columns, which goes through the cell manager:
- store: https://github.com/CeciliaZ030/zkevm-circuits/blob/fa3ac00f3e83deecf31d991bc5d75fde4d941637/zkevm-circuits/src/circuit_tools/memory.rs#L223-L227
- load: https://github.com/CeciliaZ030/zkevm-circuits/blob/fa3ac00f3e83deecf31d991bc5d75fde4d941637/zkevm-circuits/src/circuit_tools/memory.rs#L239-L243

Without different tags, the cell manager could store these values in any of the two columns, but they really need to be stored in separate columns as the lookup expects.

CeciliaZ030 commented on zkevm-circuits/src/circuit_tools/constraint_builder.rs:
```diff
@@ -1177,184 +1071,62 @@ macro_rules! _require {
             " => @",
             stringify!($tag),
         );
         $cb.add_lookup(
```
Much of the flags are removed, I'm sad but it does become more user friendly based on Yue's feedback. I still suggest preserving a RLC flag that just combine all the `$values` and `$table`, since RLC is a common usage and only one flag would not be too confusing.

Brechtpd replied:
Lookups are now "raw" lookups, as in the input/table expressions will be used by halo2 pretty much directly. So not sure how rlc encoding with the current `add_lookup` behavior would help much because if you need RLC encoding you can just do it on the inputs of this function (which is what the lookups generated in `add_columns` in the cell manager currently does). I think this also makes sense for users so that `add_lookup` is easy to understand and the values used in the inputs can be specified directly by the user I think still in an easy way (special RLC option vs just doing `rlc::expr` on the values like they are used to with no internal magic).

CeciliaZ030 commented:
In order to sell this SDK to developers in Singapore, these changes are great! After we merge this I'm gonna make PR with some new minor changes to backend that I have during PI/Anchor, and then I'll use this version for the standalone repo :)

CeciliaZ030 commented:
I migrated my new changes coming from making PI/Anchor. Some of them like storing equality constraints in CB and building later is because PI use equality constraints, so I just added this API for potential user case. 

Another thing is I think there should be a struct that convey information from conig stage to assignment stage thus I introduced `AssignParams`. It will reduce the memory footprint cuz the object CB would be dropped after configuration and only what's needed are passed on. In future iteration I might add more thing to this `AssignParams` mainly for annotations (we can pass information out like cell name, value, ... then annotate with cached region APIs), and for now I also added a annotate columns function using CellType as names.

Brechtpd commented on zkevm-circuits/src/mpt_circuit.rs:
```diff
@@ -238,192 +226,136 @@ impl<F: Field> MPTConfig<F> {
             });
-
-            cb.base.build_constraints()
+            cb.base.build_constraints_selected(meta.query_fixed(q_enable, Rotation::cur()))
```
I think I said it before, but I don't like this because this moves the condition to after you've read all the constraints which is very confusing, because you have no idea when the constraints are active because you only know it when you get to the end of the block of code. So you read the constraints, but don't know what `if` statement they belong to. 

It's also something that has no equivalence to normal programming language (except do while loops, but nobody uses because they are crazy!). If you program in rust/c++/typescript/python, would you ever want to move the condition of an `if` block to the end? If the code inside an `if` is too big it should be moved inside a separate function.

If users really want to (but I really don't think they should), it's also possible to achieve this without an `ifx` statement by just doing `cb.push_condition(f!(q_enable));` at the start of the block, so no special function is needed.

Brechtpd reviewed (commented):
> Another thing is I think there should be a struct that convey information from conig stage to assignment stage thus I introduced AssignParams. It will reduce the memory footprint cuz the object CB would be dropped after configuration and only what's needed are passed on. In future iteration I might add more thing to this AssignParams mainly for annotations (we can pass information out like cell name, value, ... then annotate with cached region APIs), and for now I also added a annotate columns function using CellType as names.

The data in the ConstraintBuilder/CellManager is tiny though, so I don't think this is necessary and mainly just adds complexity. It's just the constraints/cells per row, so the amount of data this takes is very small. Especially compared to the memory requirements for the circuits synthesize because that will actually take up memory for ALL the rows and columns using 32 bytes per cell. Memory requirements for the configure phase is maybe megabytes, while for synthesize it will be gigabytes.

ryanycw committed:
feat: rust-doc to github pages (#1578)

### Description

This pull request aims to run the cargo doc in CI to generate and
publish the doc to GitHub Pages.

### Issue Link

Solving #1206 

### Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing
functionality to not work as expected)
- [ ] This change requires a documentation update

### Contents

The implementation is in _.github/workflows/github-pages.yml_.
The goal is to generate the doc, upload the doc, and deploy the page.

### How Has This Been Tested?

I have tested this in my [forked
repository](https://github.com/ryanycw/zkevm-circuits/actions/workflows/github-pages.yml).
And here is the
[demo](https://ryanycw.github.io/zkevm-circuits/zkevm_circuits/) of what
the result will look like.

---------

Co-authored-by: Chih Cheng Liang <chihchengliang@gmail.com>

ChihChengLiang committed:
use StepRws to make rw index less error prone (#1582)

### Description

Manually counting read-write indices is error-prone. We use the
`StepRws` struct to automatically track read-write indices.

This PR is motivated by trying to fix #1565. Unfortunately, this PR
alone doesn't fix #1565.


### Issue Link


### Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing
functionality to not work as expected)
- [ ] This change requires a documentation update

lightsing committed:
[fix] create coinbase account in EndTx (#1583)

### Description

This PR adds additional logic to the EVM circuit to handle empty
coinbase accounts, including reading and writing of their code hash and
checking if the previous coinbase code hash is zero.

### Issue Link


https://github.com/privacy-scaling-explorations/zkevm-circuits/issues/1554

### Type of change

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing
functionality to not work as expected)
- [ ] This change requires a documentation update

### Contents

- Added reading and writing of coinbase code hash in
`bus-mapping/src/evm/opcodes/begin_end_tx.rs`
- Modified the logic of EVM circuit `end_tx.rs` to handle empty coinbase
account
- Update the `TransferToGadget` struct with additional fields and
methods

### Rationale

The additional logic ensures that the EVM circuit accounts for empty
coinbase accounts and properly handles updates to their state.

### How Has This Been Tested?

---------

Co-authored-by: DreamWuGit <wwuwwei@126.com>

CeciliaZ030 commented:
Reverted some changes.

KimiWu123 committed:
Feat/#1519 implement ErrorOutOfGasDynamicMemoryExpansion error state (#1567)

### Description

This PR was pulled from Scroll team
(https://github.com/scroll-tech/zkevm-circuits/pull/351) and
1. moved CREATE case to a sparated gadget, see
https://github.com/privacy-scaling-explorations/zkevm-circuits/pull/1562.
2. supported uint64 overflow
3. updated the rlc part to fit our word lo/hi design.

### Issue Link

closed #1519 

### Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing
functionality to not work as expected)
- [ ] This change requires a documentation update

---------

Co-authored-by: Akase Cho <lightsing@users.noreply.github.com>

grandchildrice committed:
remove return_revert comment (#1587)

### Description

I removed the TODO comment on return_revert.
As @ChihChengLiang mentioned in the issue, that has already been solved
on
https://github.com/privacy-scaling-explorations/zkevm-circuits/pull/908
.

### Issue Link


https://github.com/privacy-scaling-explorations/zkevm-circuits/issues/1194

### Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing
functionality to not work as expected)
- [ ] This change requires a documentation update

### Contents

- remove the comment

Brechtpd committed:
Add minimal RLP encoding check for RLP values

Brechtpd committed:
Feedback
